src/sidebar/search-client.js

```diff
@@ -122,15 +126,15 @@ export default class SearchClient extends TinyEmitter {
         this._results = this._results.concat(chunk);
       }
 
-      // Check if there are additional pages of results to fetch. In addition to
-      // checking the `total` figure from the server, we also require that at
-      // least one result was returned in the current page, otherwise we would
-      // end up repeating the same query for the next page. If the server's
-      // `total` count is incorrect for any reason, that will lead to the client
-      // polling the server indefinitely.
-      const nextOffset = offset + results.rows.length;
-      if (results.total > nextOffset && chunk.length > 0) {
-        this._getBatch(query, nextOffset);
+      // If the current batch was full, there might be additional batches available.
+      const nextBatchAvailable = chunk.length === this._chunkSize;
+
+      // Get the cursor for the start of the next batch.
+      const nextSearchAfter =
+        chunk.length > 0 ? chunk[chunk.length - 1][this._sortBy] : null;
+
+      if (nextBatchAvailable && nextSearchAfter) {
+        this._getBatch(query, nextSearchAfter);
```

Comment:
Note that we no longer need to worry about the `total` figure returned by the backend being wrong, which we observed could happen in the past [1]. Instead we just keep fetching pages of results until we get one that is not full.

[1] I can't recall if this was due to Elasticsearch's result ("hits") count not being completely accurate in some circumstances or if there was some other cause, but we had needed to code around it in the client.

Reply:
If for example the client is fetching pages of 200 results and there are exactly 600 results then the client will fetch 3 full pages of results and then request a 4th page (since the previous page was full), which comes back empty. At that point the search stops.